File: clients/python/solana/status.py
"""
solana/status.py -- Read campaign state from Solana (no transaction).

Usage:
    python -m clients.python sol:status
"""

from __future__ import annotations

import argparse
import asyncio
import json
import sys

from clients.python.solana import config as sol_config
from clients.python.solana.client import get_client, load_keypair, load_idl
from clients.python.shared.output import _now_iso, print_error


async def _status(variant: str) -> dict:
    """Read and return campaign state as a dict."""
    from anchorpy import Program, Provider, Wallet
    from solders.pubkey import Pubkey

    py_idl_path, program_id_str = sol_config.SOLANA_VARIANT_ARTIFACTS[variant]
    program_id = Pubkey.from_string(program_id_str)

    client = get_client()
    payer = load_keypair()
    wallet = Wallet(payer)
    provider = Provider(client, wallet)

    idl = load_idl(py_idl_path)
    program = Program(idl, program_id, provider)

    campaign_addr = sol_config.SOLANA_CAMPAIGN_PDA
    if not campaign_addr:
        raise EnvironmentError("SOLANA_CAMPAIGN_ADDRESS env var is not set.")
    campaign_pda = Pubkey.from_string(campaign_addr)

    # Fetch the campaign account data
    campaign_data = await program.account["Campaign"].fetch(campaign_pda)

    await client.close()

    # Convert to dict -- anchorpy returns a namespace object
    return {
        "chain": "solana",
        "operation": "status",
        "variant": variant,
        "campaign": str(campaign_pda),
        "creator": str(campaign_data.creator) if hasattr(campaign_data, 'creator') else None,
        "totalRaised": campaign_data.total_raised if hasattr(campaign_data, 'total_raised') else None,
        "finalized": campaign_data.finalized if hasattr(campaign_data, 'finalized') else None,
        "softCap": campaign_data.soft_cap if hasattr(campaign_data, 'soft_cap') else None,
        "hardCap": campaign_data.hard_cap if hasattr(campaign_data, 'hard_cap') else None,
        "deadline": campaign_data.deadline if hasattr(campaign_data, 'deadline') else None,
        "currentMilestone": campaign_data.current_milestone if hasattr(campaign_data, 'current_milestone') else None,
        "timestamp": _now_iso(),
    }


def main(args=None) -> None:
    parser = argparse.ArgumentParser(description="Read Solana campaign state.")
    parser.add_argument("--variant", default=sol_config.SOLANA_VARIANT)
    parsed = parser.parse_args(args)

    try:
        result = asyncio.run(_status(parsed.variant))
        print(json.dumps(result))
    except Exception as exc:
        print_error("status", exc, chain="solana")


if __name__ == "__main__":
    main()

File: clients/ts/src/solana/status.ts
import { parseArgs } from "node:util";
import { PublicKey } from "@solana/web3.js";
import {
  connection,
  wallet,
  program,
} from "./config.js";
import { contributorRecordPda, resolveCampaign } from "./pda.js";
import { printResult, printError } from "../shared/output.js";

const { values } = parseArgs({
  options: {
    campaign: { type: "string" },
    contributor: { type: "string" },
  },
  strict: false,
});

async function main() {
  const start = performance.now();

  const campaignAddr = resolveCampaign(values["campaign"], wallet.publicKey);

  const campaignAccount = await (program.account as any).campaign.fetch(campaignAddr);

  const data: Record<string, unknown> = {
    campaignAddress: campaignAddr.toBase58(),
    creator: campaignAccount.creator.toBase58(),
    paymentMint: campaignAccount.paymentMint.toBase58(),
    receiptMint: campaignAccount.receiptMint.toBase58(),
    softCap: campaignAccount.softCap.toString(),
    hardCap: campaignAccount.hardCap.toString(),
    deadline: campaignAccount.deadline.toString(),
    totalRaised: campaignAccount.totalRaised.toString(),
    totalWithdrawn: campaignAccount.totalWithdrawn.toString(),
    finalized: campaignAccount.finalized,
    successful: campaignAccount.successful,
    currentMilestone: campaignAccount.currentMilestone,
    milestoneCount: campaignAccount.milestoneCount,
    milestones: Array.from(campaignAccount.milestones).slice(0, campaignAccount.milestoneCount),
    campaignId: campaignAccount.campaignId.toString(),
  };

  // Optional: read contributor record
  const contributorAddr = values["contributor"] ?? wallet.publicKey.toBase58();
  try {
    const recordPda = contributorRecordPda(campaignAddr, new PublicKey(contributorAddr));
    const record = await (program.account as any).contributorRecord.fetch(recordPda);
    data.contribution = record.amount.toString();
    data.contributorAddress = contributorAddr;
  } catch {
    data.contribution = "0";
    data.contributorAddress = contributorAddr;
  }

  const elapsed = performance.now() - start;

  printResult({
    chain: "solana",
    operation: "status",
    txHash: null,
    blockNumber: null,
    gasUsed: null,
    status: "success",
    timestamp: new Date().toISOString(),
    elapsedMs: Math.round(elapsed),
    data,
  });
}

main().catch((err) => printError("status", err, "solana"));
